src/GovUk.Education.ExploreEducationStatistics.Public.Data.Processor/Services/DataSetVersionChangeService.cs

```diff
@@ -0,0 +1,425 @@
+
+namespace GovUk.Education.ExploreEducationStatistics.Public.Data.Processor.Services;
+
+internal class DataSetVersionChangeService(PublicDataDbContext publicDataDbContext)
+    : IDataSetVersionChangeService
+{
+    public async Task CreateChanges(
+        Guid nextDataSetVersionId,
+        CancellationToken cancellationToken = default)
+    {
+        var nextVersion = await publicDataDbContext
+            .DataSetVersions
+            .Include(dsv => dsv.DataSet)
+            .SingleAsync(dsv => dsv.Id == nextDataSetVersionId, cancellationToken);
+
+        var previousVersionId = nextVersion.DataSet.LatestLiveVersionId!.Value;
+
+        var mappings = await publicDataDbContext
+            .DataSetVersionMappings
+            .AsNoTracking()
+            .SingleAsync(dsvm => dsvm.TargetDataSetVersionId == nextDataSetVersionId, cancellationToken);
+
+        (nextVersion.FilterMetaChanges, nextVersion.FilterOptionMetaChanges) = await CreateFilterChanges(
+            previousVersionId: previousVersionId,
+            nextVersionId: nextVersion.Id,
+            mappings: mappings,
+            cancellationToken: cancellationToken);
+
+        (nextVersion.LocationMetaChanges, nextVersion.LocationOptionMetaChanges) = await CreateLocationChanges(
+            previousVersionId: previousVersionId,
+            nextVersionId: nextVersion.Id,
+            mappings: mappings,
+            cancellationToken: cancellationToken);
+
+        nextVersion.GeographicLevelMetaChange = await CreateGeographicLevelChange(
+            previousVersionId: previousVersionId,
+            nextVersionId: nextVersion.Id,
+            cancellationToken: cancellationToken);
+
+        nextVersion.IndicatorMetaChanges = await CreateIndicatorChanges(
+            previousVersionId: previousVersionId,
+            nextVersionId: nextVersion.Id,
+            cancellationToken: cancellationToken);
+
+        nextVersion.TimePeriodMetaChanges = await CreateTimePeriodChanges(
+            previousVersionId: previousVersionId,
+            nextVersionId: nextVersion.Id,
+            cancellationToken: cancellationToken);
+
+        await publicDataDbContext.SaveChangesAsync(cancellationToken);
+    }
+
+    private async Task<(List<FilterMetaChange> filterMetaChanges, List<FilterOptionMetaChange> filterOptionMetaChanges)>
+        CreateFilterChanges(
+            Guid previousVersionId,
+            Guid nextVersionId,
+            DataSetVersionMapping mappings,
+            CancellationToken cancellationToken)
+    {
+        var oldFilterMetas = await GetFilterMetas(previousVersionId, cancellationToken);
+        var newFilterMetas = await GetFilterMetas(nextVersionId, cancellationToken);
+
+        var filterMetaDeletionsAndChanges = mappings.FilterMappingPlan.Mappings
+            // DO NOT INCLUDE CHANGES FOR KEYS THAT HAVE BEEN MAPPED TO THE EXACT SAME CANDIDATE KEY (NO RENAMING)
+            .Where(kv => kv.Key != kv.Value.CandidateKey)
```

Comment:
I haven't thought about this for over 3 weeks now so I may be misremembering, and I may have also misinterpreted how it all works when I first did it. However, **I think** my thoughts on this were:

- if something has changed, be it the label, hint, or whatever, then there will be a new entry in the DB for that new meta
- when we are mapping, we map 'keys' to 'candidate keys'. And we have some generators in `MappingKeyFunctions` which can be used to get the right key for each meta. For example, for `FilterMeta` we use the `PublicId`, but for `FilterOptionMeta` we use the `Label`
- in order to make a change, we **have** to be able to create a mapping for it
- the only way we can map something, is by using the keys, as mentioned before. So these keys must be the source of truth for detecting changes


If I've made a mistake, perhaps it's that I have assumed if a `FilterMeta` has changed in any way, then it's `PublicId` must have been updated. And if a `FilterOptionMeta` has changed, it must have had it's `Label` changed, and therefore it's mapping key.

Reply:
Yeh, might be a bit of a misunderstanding here I think!

The mapping plan is primarily here to resolve cases where we can't automatically determine if a meta type's 'identity' was changed or if it was deleted.

The 'identity' of a meta type is basically the mapping key dictated by the mapping key generators. For filters, the 'identity' are the filter column in the CSV (aka public IDs).

However, we're overlooking that there are automatic updates that are performed whenever mapping keys don't change. In the case of filters, other properties like label and hint can change without directly affecting the meta's identity. These **do not** require manual intervention and don't participate in the mapping process.

Consequently, I don't think that using just the mapping plan is currently sufficient and leans things more heavily to the other approach I outlined in https://github.com/dfe-analytical-services/explore-education-statistics/pull/5227#discussion_r1751765238.

Note that for filter options, the identity is the label and we currently don't allow these to be changed in any other way. However, as mentioned in my other comment, we may allow analysts to configure the 'aggregate' option in the future, but this would still be an automatic update as long as the identity can be determined (i.e. the label hasn't changed).

Reply:
This has now been refactored to allow for 'silent' updates to the metadata, whereby the Public ID hasn't changed but other fields have, such as the column name. The same has been done for locations and indicators